### PR TITLE
ddl/ingest: refactor checkpoint manager

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: ref #54436

Problem Summary:

### What changed and how does it work?

- Move `FinishAndUnregisterEngines()` out of `AddIndexIngestPipeline`. Thus, we put `Register()` and `Unregister()` to the same code level.
- Add an option to `FinishAndUnregisterEngines()` to control the behavior of "clean up local path data" and "check duplicated entry" in different scenarios.
- Refine logs.
- Remove `BackendCtx.SetDone()` and `BackendCtx.Done()`, use `adjustStartKey` to skip imported data.
- Simplify checkpoint manager logic. Now we use `importedKeyLowWatermark` to determine `startKey` in next round(see `getCheckpointReorgHandle`).
- Add integration tests for checkpoint manager.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/ddl/index.go [hunk 1]
```diff
@@ -954,28 +954,6 @@ func runIngestReorgJobDist(w *worker, d *ddlCtx, t *meta.Meta, job *model.Job,
...
-		if indexInfo.Unique {
-			hasUnique = true
-			break
-		}
-	}
-	bc, err = ingest.LitBackCtxMgr.Register(ctx, job.ID, hasUnique, nil, discovery, job.ReorgMeta.ResourceGroupName)
-	if err != nil {
-		ver, err = convertAddIdxJob2RollbackJob(d, t, job, tbl.Meta(), allIndexInfos, err)
-		return false, ver, errors.Trace(err)
-	}
```
Comment (tangenta) [inline pkg/ddl/index.go:978]: Will be registered in `runAddIndexInLocalIngestMode`.

File: pkg/ddl/ingest/integration_test.go [hunk 2]
```diff
@@ -379,3 +380,51 @@ func TestMultiSchemaAddIndexMerge(t *testing.T) {
...
+		rowCnt.Add(1)
+	})
+	tk.MustExec("alter table t add index idx(b);")
+	require.True(t, injected)
+	// Write conflict error should not retry the whole job.
+	require.Equal(t, 3, int(rowCnt.Load())) // it should not be 6
+	tk.MustExec("admin check table t;")
+}
+
+func TestAddIndexIngestPartitionCheckpoint(t *testing.T) {
```
Comment (lance6716) [inline pkg/ddl/ingest/integration_test.go:417]: do we have something to check in this test?

**Review by lance6716 — COMMENTED** [review]
7 / 17 files viewed

File: pkg/ddl/ingest/engine_mgr.go [hunk 3]
```diff
@@ -101,8 +101,20 @@ func (bc *litBackendCtx) Register(indexIDs []int64, uniques []bool, tbl table.Ta
 	return ret, nil
 }
 
+// UnregisterOpt controls the behavior of backend context unregistering.
+type UnregisterOpt int
+
+const (
+	// OptBasic only closes engines, it does not clean up sort path data.
+	OptBasic UnregisterOpt = iota << 1
```
Comment (D3Hunter) [inline pkg/ddl/ingest/engine_mgr.go:109]:
```suggestion
	OptCloseEngines UnregisterOpt = iota << 1
```

Comment (D3Hunter) [inline pkg/ddl/ingest/engine_mgr.go:109]:
```suggestion
	OptBasic UnregisterOpt = 1 << iota
```

**Review by D3Hunter — COMMENTED** [review]
rest lgtm

File: pkg/ddl/ingest/integration_test.go (hunk 2 above)
Comment (tangenta) [reply pkg/ddl/ingest/integration_test.go:417]: https://github.com/pingcap/tidb/pull/54747/commits/eaa11ed4af79fa91bc6fce6548b879becb4f2549

File: pkg/ddl/ingest/engine_mgr.go (hunk 3 above)
Comment (tangenta) [reply pkg/ddl/ingest/engine_mgr.go]: https://github.com/pingcap/tidb/pull/54747/commits/eaa11ed4af79fa91bc6fce6548b879becb4f2549

Comment (tangenta) [reply pkg/ddl/ingest/engine_mgr.go]: https://github.com/pingcap/tidb/pull/54747/commits/eaa11ed4af79fa91bc6fce6548b879becb4f2549

**Review by D3Hunter — APPROVED** [review]

File: pkg/ddl/job_table.go [hunk 4]
```diff
@@ -776,38 +776,30 @@ func getDDLReorgHandle(se *sess.Session, job *model.Job) (element *meta.Element,
 	return
 }
 
-func getCheckpointReorgHandle(se *sess.Session, job *model.Job) (startKey, endKey kv.Key, physicalTableID int64, err error) {
-	startKey, endKey = kv.Key{}, kv.Key{}
+func getCheckpointReorgHandle(se *sess.Session, job *model.Job) (adjustedStart kv.Key, physicalTableID int64, err error) {
```
Comment (lance6716) [inline pkg/ddl/job_table.go:779]: we don't "adjust" the start key? maybe keep the name startKey

**Review by lance6716 — APPROVED** [review]

**Pull request merged**